runtime/src/stake_account.rs

```diff
@@ -0,0 +1,72 @@
+#[cfg(RUSTC_WITH_SPECIALIZATION)]
+use solana_frozen_abi::abi_example::AbiExample;
+use {
+    solana_sdk::{
+        account::{AccountSharedData, ReadableAccount},
+        account_utils::StateMut,
+        instruction::InstructionError,
+        pubkey::Pubkey,
+        stake::state::{Delegation, StakeState},
+    },
+    thiserror::Error,
+};
+
+#[derive(Clone, Debug, Default, PartialEq)]
+pub struct StakeAccount(AccountSharedData, StakeState);
```

Comment:
The `Stakes` struct ensures that only delegations are cached; but as for `StakeAccount`, I would rather keep it more generic and oblivious to delegations for now so that this has future compatibility and potentially can be used elsewhere in the code.

For one example, if I restrict `StakeAccount` to delegations, I wouldn't be able to replace `(StakeState, AccountSharedData)` in `VoteWithStakeDelegations` with `StakeAccount`:
https://github.com/solana-labs/solana/blob/aea17c35a/runtime/src/bank.rs#L1263
because currently when a stake account is loaded, we do not check if it is indeed a delegation or not:
https://github.com/solana-labs/solana/blob/aea17c35a/runtime/src/bank.rs#L2615-L2637

Reply:
> The Stakes struct ensures that only delegations are cached; but as for StakeAccount, I would rather keep it more generic and oblivious to delegations for now so that this has future compatibility and potentially can be used elsewhere in the code.

Good point, I suggest renaming it to `StakeDelegationAccount` to avoid confusion.

> because currently when a stake account is loaded, we do not check if it is indeed a delegation or not:

Good point, do you think you can add that check as part of this PR behind the same feature gate? If not, I can help do it in a follow up.

Reply:
I do see some value in keeping `StakeAccount` generic and oblivious to delegations for now so that this has future compatibility and potentially can be used elsewhere in the code. Also rather not change the old code path.
Any objections to keeping this as is for now and revisit later once the old code can be removed?

Reply:
Yeah I can see value in it too but we don't have an immediate usecase for that so I don't think the maintenance burden is worth. If we do need this to be more generic in the future, there are a lot of new `unwrap` calls that could break. I think if we make it type-safe now, it will be easier to make future changes.

You raise a great point about not changing the old code path. I was concerned that my suggested change wouldn't work with how we build `VoteStakeDelegations` so I tried to implement it to see if it was workable. I have opened a PR with that implementation here: https://github.com/behzadnouri/solana/pull/49

Reply:
Implemented some type-safety using phantom-data.
`Stakes` struct now uses `StakeAccount<Delegation>` which can only wrap a stake-state which is a delegation and so it implements
```rust
fn delegation(&self) -> Delegation
```
instead of `-> Option<Delegation>`, so removed `unwrap`s.

But `VoteWithStakeDelegations` uses `StakeAccount<()>` which wraps any account with stake state and does not implement any `fn delegation(&self) -> ...` methods.

Reply:
Nice solution, that works for me!